code/implementation/host-stats/src/main/java/io/cattle/platform/host/stats/api/ServiceContainerStatsLinkHandler.java

```diff
@@ -0,0 +1,117 @@
+
+import io.cattle.platform.api.link.LinkHandler;
+import io.cattle.platform.core.constants.InstanceConstants;
+import io.cattle.platform.core.model.Host;
+import io.cattle.platform.core.model.Instance;
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.docker.util.DockerUtils;
+import io.cattle.platform.host.model.HostApiAccess;
+import io.cattle.platform.host.service.HostApiService;
+import io.cattle.platform.host.stats.utils.StatsConstants;
+import io.cattle.platform.object.ObjectManager;
+import io.cattle.platform.token.TokenService;
+import io.github.ibuildthecloud.gdapi.context.ApiContext;
+import io.github.ibuildthecloud.gdapi.request.ApiRequest;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+public class ServiceContainerStatsLinkHandler implements LinkHandler {
+
+    HostApiService hostApiService;
+    ObjectManager objectManager;
+    TokenService tokenService;
+
+    private static final String SERVICE_PATH = "/service/";
+
+    @Override
+    public String[] getTypes() {
+        return new String[] { "service" };
+    }
+
+    @Override
+    public boolean handles(String type, String id, String link, ApiRequest request) {
+        return StatsConstants.CONTAINER_STATS.equalsIgnoreCase(link);
+    }
+
+    @Override
+    public Object link(String name, Object obj, ApiRequest request) throws IOException {
+        Service service = (Service) obj;
+
+        List<Instance> serviceInstances = objectManager.mappedChildren(service, Instance.class);
+
+        if (service == null) {
+            return null;
+        }
+
+        String metaUrl = null;
+        StatsAccess meta = new StatsAccess();
+        List<StatsAccess> serviceStatsQuery = new ArrayList<>();
+        for (Instance instance : serviceInstances) {
+            if (!InstanceConstants.STATE_RUNNING.equals(instance.getState()) && !InstanceConstants.STATE_STOPPED.equals(instance.getState())) {
+                continue;
+            }
+            Host host = DockerUtils.getHostFromContainer(objectManager, instance);
+            Map<String, Object> payload = new HashMap<>();
+            Map<String, Object> containerIdsMap = new HashMap<String, Object>();
+            containerIdsMap.put(DockerUtils.getDockerIdentifier(instance),
+                    ApiContext.getContext().getIdFormatter().formatId(objectManager.getType(instance), instance.getId()));
+            payload.put("containerIds", containerIdsMap);
+            HostApiAccess apiAccess = hostApiService.getAccess(request, host.getId(), payload, new String[] { StatsConstants.CONTAINER_STATS_PATH.get() });
+            if (apiAccess == null) {
+                continue;
+            }
+
+            StringBuilder url = new StringBuilder();
+            
+            url.append(apiAccess.getUrl());
+
+            metaUrl = url.toString() + SERVICE_PATH;
```

Comment:
Nit: as url is already a builder, you can just use append to concat SERVICE_PATH to it, and then call toString later